### PR TITLE
counter: Update counter API in order to provide more flexibility

### PR DESCRIPTION
**Pull request by pizi-nordic**
This PR introduces new "reset" parameter to the set_top_value()
making resetting counter optional during change of top value.

Such change allows for #12068 implementation on hardware which
does not provide alarms.

### CODE REVIEW

File: include/counter.h
Comment (nordic-krch) [inline include/counter.h]: add -ENOTSUP error if option is not supported

File: include/counter.h
Comment (nordic-krch) [inline include/counter.h]: i wonder if it's not worth to bundle options to cfg structure like in `counter_set_channel_alarm`. It's then easier to add new features without braking backward compatibility (as long as false in option means legacy behavior). Now we exceed boundary of 4 arguments.

**Review by nordic-krch — CHANGES_REQUESTED** [review]

File: include/counter.h
Comment (pizi-nordic) [reply include/counter.h]: I think that with 5 arguments we are still more efficient that using separate data structure.

File: include/counter.h
Comment (pizi-nordic) [reply include/counter.h]: Updated.

**Comment by pizi-nordic** [comment]
@nordic-krch: Could you please look at the last update?

**Review by nordic-krch — APPROVED** [review]

File: include/counter.h
Comment (ioannisg) [inline include/counter.h]: counter***

File: drivers/counter/timer_dtmr_cmsdk_apb.c
Comment (ioannisg) [inline drivers/counter/timer_dtmr_cmsdk_apb.c]: I think some inline comment is needed here, regarding the function of the background load

File: include/counter.h
Comment (pizi-nordic) [reply include/counter.h]: Fixed. Thank you.

File: drivers/counter/timer_dtmr_cmsdk_apb.c
Comment (pizi-nordic) [reply drivers/counter/timer_dtmr_cmsdk_apb.c]: Added.

Comment (ioannisg) [reply drivers/counter/timer_dtmr_cmsdk_apb.c]: thanks

**Review by ioannisg — APPROVED** [review]
Looks ok to me. Does the counter test still pass with the modified API?

**Comment by pizi-nordic** [comment]
@ioannisg: As seen in the shippable.

**Comment by ioannisg** [comment]
> 
> 
> @ioannisg: As seen in the shippable.

Does Shippable actually run this test on target?

**Review by anangl — COMMENTED** [review]
The idea of adding the `reset` parameter is good, but I think it would be indeed better to use a structure for passing the parameters to the `counter_set_top_value` function, as @nordic-krch [suggested earlier](https://github.com/zephyrproject-rtos/zephyr/pull/14794/files#r268050055). In most cases it will be more efficient than passing 5 arguments to the function (4 via registers and the fifth via stack), especially when the contents of the structure will be constant, what I think can be expected to happen often. And we'll gain the flexibility mentioned by @nordic-krch. Please, consider it once more. The amount of work needed to switch to this approach does not seem to be that big. I can help a bit if needed.

**Comment by nordic-krch** [comment]
@pizi-nordic @anangl can you take a look as I've updated it to use configuration struct instead of function arguments.

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (pizi-nordic) [inline drivers/counter/counter_ll_stm32_rtc.c]: ARG_UNUSED(cfg) ?

File: drivers/counter/counter_rtc_qmsi.c
Comment (pizi-nordic) [inline drivers/counter/counter_rtc_qmsi.c]: ARG_UNUSED(cfg)?

File: include/counter.h
Comment (anangl) [inline include/counter.h]: This should be `const struct counter_top_cfg *cfg`. Drivers are not supposed to change the contents of this structure, right?

File: include/counter.h
Comment (pizi-nordic) [inline include/counter.h]: ... Cannot be NULL.

**Comment by nordic-krch** [comment]
@pizi-nordic @anangl thanks for swift response. Even swifter reaction from my side. Suggestions applied.

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (anangl) [inline drivers/counter/counter_ll_stm32_rtc.c]: This parameter is still unused in this function.

File: drivers/counter/counter_rtc_qmsi.c
Comment (anangl) [inline drivers/counter/counter_rtc_qmsi.c]: Same story here.

File: include/counter.h
Comment (anangl) [inline include/counter.h]: Shouldn't this be the first field? It's the main one, and not optional. I know, I'm picky.

Comment (nordic-krch) [reply include/counter.h]: wanted to make if similar as the one above:
```
struct counter_alarm_cfg {
	counter_alarm_callback_t callback;
	u32_t ticks;
	void *user_data;
	bool absolute;
};
```

will flip it.

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_ll_stm32_rtc.c]: used here: `const struct counter_config_info *info = dev->config->config_info;`

File: drivers/counter/counter_rtc_qmsi.c
Comment (nordic-krch) [reply drivers/counter/counter_rtc_qmsi.c]: same story here 😄

Comment (anangl) [reply drivers/counter/counter_rtc_qmsi.c]: Right. My bad, in both stories. 😄
How could I not see this right above the line I commented? I'm getting old, I guess...

**Review by anangl — APPROVED** [review]

**Comment by nordic-krch** [comment]
@mnkp @MaureenHelm @pabigot  can you take a look as well?

**Comment by mnkp** [comment]
I believe I miss some information. The way I see it: if we want to trigger a callback when the counter hits some value we call it an alarm and use `counter_set_channel_alarm` function. If we want to trigger a callback when the counter hits some value and reset the counter we call it the top value and use `counter_set_top_value` function. So what does it mean that the hardware does not provide alarms? Shouldn't we rather state that the hardware doesn't allow setting the top value?

**Comment by nordic-krch** [comment]
@mnkp PR is about resetting counter when setting the top value and not about resetting when top value is reached. Current API states that when you change top value counter is reset. In some cases you want to keep it free running and not reset it while changing top value and this patch allows user to specify that.

**Comment by mnkp** [comment]
> PR is about resetting counter when setting the top value and not about resetting when top value is reached.

@nordic-krch Thanks, that's clear now.

Taking into account that the `counter_set_top_value` can be called when the counter is running I believe we need to better clarify one aspect of the new behavior. Let's assume the top value of the 32-bit counter is configured as 2000. Then the user calls `counter_set_top_value` and sets the top value to 1000 at the time when the counter's value is 1500.

Should the function return an error in this case, should we reset the counter even if the `reset` parameter is set to false, should we trigger a callback? Looking at the current implementation it seems that we simply update the counter's top value and let it reload when it hits the max top value. That may be a long wait (like in the above example for the 32-bit counter) and not what the user expects. We need to document better the intended behavior. Also update the testcase to cover the aforementioned scenario.

**Comment by nordic-krch** [comment]
@mnkp meanwhile I've #16233 which will enable to add counter device capabilities. That will help to determine on runtime device capabilities and use that for testing.

Regarding:
> That may be a long wait (like in the above example for the 32-bit counter) and not what the user expects.

I would leave it to the user which is must use this feature wisely rather than adding logic to the driver.

**Comment by mnkp** [comment]
> > That may be a long wait (like in the above example for the 32-bit counter) and not what the user expects.
> 
> I would leave it to the user which is must use this feature wisely rather than adding logic to the driver.

I don't have strong opinion on this but we need to better document expected behavior.

**Comment by nordic-krch** [comment]
@mnkp I've updated comment to explain that there is a risk of going out of bounds and that it is the user responsibility.

@pizi-nordic @anangl I've also updated `alarm_top_cfg` structure to use `flags` instead of `reset` flag. Flags are more flexible for further expansion.

File: drivers/counter/counter_nrfx_rtc.c
Comment (mnkp) [inline drivers/counter/counter_nrfx_rtc.c]: This part is a bit tricky. User of the API may want to change the top value up and down by just a few counts. I.e. a PI controller would do exactly such a thing. `nrfx_rtc_cc_disable` disables channel events and channel interrupts. If in between the call to `nrfx_rtc_cc_disable` and `nrfx_rtc_cc_set` the count value would pass the new 'top value' we would fail to trigger the callback and have 'runaway' counter.

An example: 
- user calls `counter_set_top_value` when the count value is 900
- when `nrfx_rtc_cc_disable` is finally executed the counter has value 999
- when `nrfx_rtc_cc_set` sets the 'top value' to 1000 the count value is 1001

This is similar, though not identical to my last comment. From the application point of view neither decreasing nor increasing the top value is safe here. I believe we cannot claim it's an application responsibility to mitigate the issue. Application, working in a preemptive RTOS, has limited control over timing. For the user of the API it's going to be impossible to estimate a safe margin (unless it's a very rough estimation) when the `counter_set_top_value` can be called in relation to the current counter value.

One safe way a counter driver can implement the feature is to
- block all the interrupts
- read the current counter value
- knowing the frequency of the counter verify that we have enough time to set the new top value before counter reloads
1. if there is not enough time to set the new top value: enable interrupts, wait until the counter reloads itself, set the new top value
2. if there is enough time: set the new top value, enable interrupts

Not calling the  `nrfx_rtc_cc_disable` would be another option but it doesn't solve the problem when user wants to decrease the top value by just a few counts.

The `COUNTER_TOP_FLAG_RESET` may have been added to support kernel usage scenarios described in  #12068 where the above issue may not be relevant at all. This is part of the public API however. Other users may use the API differently.

Comment (nordic-krch) [reply drivers/counter/counter_nrfx_rtc.c]: @mnkp user can do something like:

```
cfg.flags = 0; /* no reset */
counter_set_top_value(dev, &cfg);
if (counter_read(dev) > counter_get_top_value(dev)) {
    cfg.flags = COUNTER_TOP_FLAG_RESET;
    counter_set_top_value(dev, &cfg);
}
```
Of course, such counter will lose some ticks but that is inevitable if we missed setting it.

Alternatively, it can be inside `counter_set_top_value` and return error if counter exceeded top value.

Comment (nordic-krch) [reply drivers/counter/counter_nrfx_rtc.c]: @mnkp any comments? do  you agree that it can stay as the user responsibility?

**Comment by nordic-krch** [comment]
@erwango @MaureenHelm any comments?

File: drivers/counter/counter_nrfx_rtc.c
Comment (pabigot) [reply drivers/counter/counter_nrfx_rtc.c]: This touches on areas of concern I have in #16252, i.e. no [clear model of what a counter is](https://github.com/zephyrproject-rtos/zephyr/pull/16252#issuecomment-494174959), but looking at this I can't imagine why we would expect an application to have to deal with a case where `counter_read()`  produced a value greater than `counter_get_top_value()`, no matter what was going on.  That's just not consistent with how a counter should behave, unless "top value" doesn't mean top value.

**Comment by benpicco** [comment]
`counter_sam0_tc32.c` needs an update too.

**Comment by carlescufi** [comment]
> `counter_sam0_tc32.c` needs an update too.

@nordic-krch will you take a look at this driver?

File: drivers/counter/counter_nrfx_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_rtc.c]: So I have following proposal which would be similar to #16252: there is an additional flag which instructs driver to reset counter if top value if exceeded or return error. For certain scenarios lost ticks will be accepted and then user will set flag to reset and for other it may indicate that application has performance issues and error should be handled.

File: include/counter.h
Comment (nordic-krch) [reply include/counter.h]: done

**Comment by nordic-krch** [comment]
> counter_sam0_tc32.c needs an update too.

@benpicco i've updated it but for know i added 
```
if (top_cfg->flags & COUNTER_TOP_CFG_DONT_RESET) {
    return -ENOTSUP;
}
```
please let me know if this is supported and help me to add support. Feel free to push change to PR.

**Comment by benpicco** [comment]
@Sizurka already wrote the driver with this PR in mind - all you have to do is

```
if (top_cfg->flags & COUNTER_TOP_CFG_DONT_RESET) {
    reset = false;
}
```

(or better yet, replace the `reset` value with this check, it's only used in one place.)

File: drivers/counter/counter_nrfx_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_rtc.c]: @mnkp @pabigot what do you think?

**Comment by benpicco** [comment]
please squash the commit.

**Comment by nordic-krch** [comment]
@benpicco, squashed.

**Comment by carlescufi** [comment]
@mnkp can you please take another look and approve if happy?
@andyross if you have had a chance to look at the counter driver API PRs, could you voice your opinion on this one?

**Comment by carlescufi** [comment]
@erwango @MaureenHelm can you please take a look and give feedback?

**Comment by nordic-krch** [comment]
I've added following updates:
- driver after changing top value checks if it does not go outside and return `-ETIME` if that happened.
- added `COUNTER_TOP_CFG_RESET_WHEN_LATE` flag which additionally can instruct driver to reset counter when such situation occurs.

Added implementation for 4 drivers which can support that now: 
- `counter_sam0_tc32.c` @benpicco can you check?
- `counter_gecko_rtcc.c` @mnkp  can you check?
- `counter_nrfx_timer.c` and `counter_nrfx_rtc.c`

**Comment by benpicco** [comment]
looks good to me.

**Comment by nordic-krch** [comment]
@mnkp can you take a look at the recent changes? Lets make goal to merge it in 3 months from submitting. One week left.

**Comment by nordic-krch** [comment]
@anangl could you also take a look? Since you approved there were some changes.

File: drivers/counter/counter_mcux_rtc.c
Comment (mnkp) [inline drivers/counter/counter_mcux_rtc.c]: By default `COUNTER_TOP_CFG_DONT_RESET` is not set so this if statement will evaluate to true. This is change of behavior. I didn't check the datasheet but how come this counter doesn't start counting from 0 when it's reaching the top value?

File: drivers/counter/counter_gecko_rtcc.c
Comment (mnkp) [inline drivers/counter/counter_gecko_rtcc.c]: Could we use `cfg` instead of `top_cfg` like in all other drivers. To make it more consistent.

**Review by mnkp — COMMENTED** [review]
> @mnkp can you take a look at the recent changes? Lets make goal to merge it in 3 months from submitting. One week left.

Sorry, you're right complaining. The  #16252 is introducing the concept of a guard time. I believe it would be best to reuse it. We would keep the API consistent.

File: drivers/counter/counter_gecko_rtcc.c
Comment (nordic-krch) [reply drivers/counter/counter_gecko_rtcc.c]: done

File: drivers/counter/counter_mcux_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_mcux_rtc.c]: i've digged into the manual. Counter can be reset or left free running. I've added implementation to support resetting the counter.

> This is change of behavior

well, behavior was a bug according to `counter_set_top_value` description in master:
`Function sets top value and resets the counter to 0 or top value depending on counter direction.`

Now it will be clarified.

File: drivers/counter/counter_nrfx_timer.c
Comment (anangl) [inline drivers/counter/counter_nrfx_timer.c]: It should be `>=`.

File: drivers/counter/counter_nrfx_timer.c
Comment (anangl) [inline drivers/counter/counter_nrfx_timer.c]: Actually, even if the counter is reset, we can end up with the top value set lower than the current counter value. Although it could happen only for very low top values and accordingly high latencies between clearing the counter and setting the new compare value, but maybe such unusual scenario should be handled as well?

File: drivers/counter/counter_gecko_rtcc.c
Comment (anangl) [inline drivers/counter/counter_gecko_rtcc.c]: `ticks` are now a field of the `counter_top_cfg` structure. Does this compile?

Comment (nordic-krch) [reply drivers/counter/counter_gecko_rtcc.c]: there is no test that uses that apparently. Fixed.

File: drivers/counter/counter_nrfx_timer.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c]: fixed

File: drivers/counter/counter_mcux_rtc.c
Comment (anangl) [inline drivers/counter/counter_mcux_rtc.c]: The second part of the message is no longer valid.

File: drivers/counter/counter_nrfx_timer.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c]: fixed

File: drivers/counter/counter_mcux_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_mcux_rtc.c]: fixed

File: drivers/counter/counter_nrfx_timer.c
Comment (anangl) [reply drivers/counter/counter_nrfx_timer.c]: Cool. Should we use the same approach in `counter_nrfx_rtc`, for consistency and for people having the need to set top value to `1`?

Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c]: done

File: drivers/counter/counter_nrfx_rtc.c
Comment (anangl) [inline drivers/counter/counter_nrfx_rtc.c]: This block should be removed.

File: drivers/counter/counter_nrfx_timer.c
Comment (anangl) [inline drivers/counter/counter_nrfx_timer.c]: This one as well.

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_nrfx_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_rtc.c]: done, avoid committing before saving :)

File: drivers/counter/counter_nrfx_timer.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c]: done

**Review by anangl — APPROVED** [review]

**Comment by nordic-krch** [comment]
@mnkp your comments applied. Anything more?

**Comment by nordic-krch** [comment]
@mnkp gentle ping.

File: drivers/counter/counter_nrfx_rtc.c
Comment (mnkp) [inline drivers/counter/counter_nrfx_rtc.c]: I'm not familiar with Nordic architecture but user may want to set up top value interrupt for the sake of waking up the device from sleep. I.e. even if the callback is not defined we may still want to generate the interrupt.

File: include/counter.h
Comment (mnkp) [inline include/counter.h]: > If counter is counting up and free running [...]

This is already explained in the description of `counter_set_top_value()` function. Maybe we shouldn't repeat it here?

File: include/counter.h
Comment (mnkp) [inline include/counter.h]: There is one difference in the description and implementation of `COUNTER_TOP_CFG_RESET_WHEN_LATE` and an equivalent `COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE` flag introduced in #16252. The former resets the counter and returns `-ETIME` error in case late setting is detected the latter doesn't. The behavior should probably be aligned.

**Review by mnkp — COMMENTED** [review]
LGTM with minor comments.

File: include/counter.h
Comment (nordic-krch) [reply include/counter.h]: agree. I lean towards returning error then. User may want system to continue to work but at the same time have trace of this rather not expected situation.

File: include/counter.h
Comment (nordic-krch) [reply include/counter.h]: right, i've removed section that overlaps with `counter_set_top_value` description.

File: drivers/counter/counter_nrfx_rtc.c
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_rtc.c]: it's rather not the case on nordic devices.

**Review by mnkp — APPROVED** [review]

**Pull request merged**
